File: calibration.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <fstream>
#include <sstream>
#include <time.h>
#include <iomanip>
#include <opencv2/opencv.hpp>

#define PAT_ROW 7			// number of rows
#define PAT_COL 11			// number of columns
#define CHESS_SIZE 30.0		// [mm]

using uint = unsigned int;
const char* path_to_log = "./log/";
const char* path_to_data = "./data/";

int main() {
	// prepare video capture
	cv::VideoCapture cap(0);
	if (!cap.isOpened()) {
		std::cerr << "Cannot open webcam device. " << std::endl;
		return EXIT_FAILURE;
	}

	// prepare window
	cv::String windowName = "Calibration";	
	cv::namedWindow(windowName, CV_WINDOW_AUTOSIZE);

	std::cout << "Press space to capture frame. " << std::endl;
	std::cout << "Press 'q' to quit capturing and begin processing. " << std::endl;

	// get date
	time_t date;
	struct tm* timeinfo;
	char now[20];
	time(&date);
	timeinfo = localtime(&date);
	strftime(now, 20, "%Y%m%d-%H%M%S", timeinfo);

	// create directory to save datas
	std::ostringstream mkdir;
	mkdir << "mkdir " << path_to_data << "/" << now;
	system(mkdir.str().c_str());

	// collect corners
	cv::Size patt_size(PAT_COL, PAT_ROW);
	std::vector<cv::Point2f> corners;
	std::vector<std::vector<cv::Point2f>> img_points;
	int key = -1;
	cv::Mat frame, gray;
	while ((key = cv::waitKey(10)) != 'q') {
		cap >> frame;
		cv::imshow(windowName, frame);
		
		if (key != ' ') {
			continue;
		}

		// find chessboard corners
		auto found = cv::findChessboardCorners(
			frame, patt_size, corners);

		if (found) {
			std::cout << "Chessboard is found. "
				<< "  Count : " << img_points.size() + 1
				<< std::endl;
		}
		else {
			std::cerr << "Not found. " << std::endl;
			continue;
		}

		// convert to subpixel resolution
		cv::cvtColor(frame, gray, CV_BGR2GRAY);
		cv::find4QuadCornerSubpix(
			gray, corners, cv::Size(3, 3));
		img_points.push_back(corners);

		// save image
		std::ostringstream filename;
		filename << path_to_data
			<< "/" << now << "/"
			<< cv::format("data%2d.bmp", img_points.size());

		cv::imwrite(filename.str(), frame);
	}

	if (img_points.size() < 1) {
		std::cerr 
			<< "The number of captured points is not enough. " 
			<< std::endl;
		return EXIT_FAILURE;
	}

	// initialize coordinate
	std::vector<cv::Point3f> object;
	for (int row = 0; row < PAT_ROW; row++) {
		for (int col = 0; col < PAT_COL; col++) {
			cv::Point3d p(
				row * CHESS_SIZE,
				col * CHESS_SIZE,
				0.0);
			object.push_back(p);
		}
	}
	std::vector<std::vector<cv::Point3f>> obj_points;
	for (uint n = 0; n < img_points.size(); n++) {
		obj_points.push_back(object);
	}
	
	// calibrate
	cv::Mat cam_mat, dist_coefs;
	std::vector<cv::Mat> rvecs, tvecs;
	cv::calibrateCamera(
		obj_points,
		img_points,
		frame.size(),
		cam_mat,		// a camera's inner parameter matrix
		dist_coefs,		// a camera's distortion coefficients
		rvecs,			// rotations of the camera at each frame
		tvecs			// translations of the camera at each frame
	);


	// format result
	std::ostringstream log_txt;
	log_txt << "Date : " << now << "\n\n";
	log_txt << "Camera parameter matrix" << "\n";
	log_txt << cam_mat << "\n";
	log_txt << "Camera distortion coefficients" << "\n";
	log_txt << dist_coefs << "\n";

	// display result
	std::cout << "\n\n" << std::endl;
	std::cout << log_txt.str() << std::endl;

	// save result
	std::ostringstream log_name;
	log_name << path_to_log << now << ".dat";
	std::ofstream log_file(log_name.str());
	log_file << log_txt.str() << std::endl;
	

	return 0;
}

File: calibration.py
import argparse
import datetime.datetime

import cv2
import numpy as np

PAT_ROW = 7			# number of rows
PAT_COL = 11		# number of columns
CHESS_SIZE = 30.0	# [mm]

path_to_log = "./log/"
path_to_data = "./data/"
date_format = "%Y%m%d-%H%M%S"
filename_format = "%d.png"

parser = argparse.ArgumentParser()
parser.add_argument("-d", help="The path to the preset data (optional). ")
args = parser.parse_args()

use_preset =  args.d is not None

cap = cv2.VideoCapture(0) if not use_preset else cv2.VideoCapture(f"{args.d}{filename_format}")
if not cap.isOpened():
    exit("Cannot open a web camera or pre-captured images. ")

window_name = "Calibration"
cv2.namedWindow(window_name, cv2.WINDOW_AUTOSIZE)

print("Press space to capture frame. ")
print("Press 'q' to quit capturing and begin processing. ")

now = datetime.now().strftime(date_format)
data_dirname = f"{path_to_data}{now}"
os.makedirs(data_dirname)

# The 3D coordinates of each corner of chessboard
obj_p = np.zeros((PAT_COL * PAT_ROW, 3), np.float32)
obj_p[:,:2] = np.mgrid[0:PAT_COL, 0:PAT_ROW].T.reshape(-1, 2)

img_pts = [] # 2D coordinates of chessboard corners
obj_pts = [] # 3D coordinates of chessboard corners
criteria = (cv2.TERM_CRITERIA_EPS + cv2.TERM_CRITERIA_MAX_ITER, 30, 0.001)
key = -1
while key != 27:
    ret, frame = cap.read()
    if not ret: break

    cv2.imshow(window_name, frame)

    if not use_preset and key != ord(' '): continue

    # Find chessboard and its corners
    found, corners = cv2.findChessboardCorners(frame, (PAT_COL, PAT_ROW))
    if found:
        print(f"Chessboard is found. Count: {len(img_pts) + 1}")
    else:
        print("Chessboard is not found. ")
        continue

    # append 3D coordinates
    obj_pts.append(obj_p)

    # Refine corners and append their 2D coordinates
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
    corners = cv2.cornerSubPix(gray, corners, (11,11), (-1,-1), criteria)
    img_pts.append(corners)

    # save if it is captured now
    if not use_preset:
        cv2.imwrite(f"{data_dirname}/frame{len(img_pts)}.png", frame)

# Check if acquired points are enough
if len(img_pts) < 1:
    exit("The number of captured points is not enough. ")

# Calibrate camera
ret, k_mat, dist_coef, rvecs, tvecs = \
        cv2.calibrateCamera(obj_pts, img_pts, gray.shape[::-1], None, None)

print(k_mat)
print(type(k_mat))
